datafusion/core/src/physical_optimizer/dist_enforcement.rs

```diff
@@ -431,9 +431,22 @@ fn reorder_aggregate_keys(
                     None
                 };
                 if let Some(partial_agg) = new_partial_agg {
+                    // Build new group expressions that correspond to the schema of partial_agg
                     let mut new_group_exprs = vec![];
-                    for idx in positions.into_iter() {
-                        new_group_exprs.push(group_by.expr()[idx].clone());
+                    for (idx, position) in positions.into_iter().enumerate() {
+                        let (group_expr, name) = group_by.expr()[position].clone();
+
+                        // Update the index of columns so that they correspond to the schema of partial_agg
+                        // rather than input_schema.
+                        let group_expr = if let Some(group_col) =
+                            group_expr.as_any().downcast_ref::<Column>()
+                        {
+                            Arc::new(Column::new(group_col.name(), idx))
+                                as Arc<dyn PhysicalExpr>
```

Comment:
I think you can do something like this to get the rust compiler to do the right cast for you:

```suggestion
                                as _
```

However the rest of this module is in the same `as Arc<...>` style so no need to change it in this PR

Reply:
Good catch!!  It is a bug.  

Regarding the fix, can we call this method to get the new `new_group_exprs` ?

`let new_group_exprs: Vec<Arc<dyn PhysicalExpr>> = partial_agg.output_group_expr();`